Classes/common/query/CDTQIndex.m

```diff
@@ -0,0 +1,171 @@
+        return nil;
+    }
+    
+    if (!indexName) {
+        LogError(@"No index name provided.");
+        return nil;
+    }
+    
+    if (!indexType) {
+        LogError(@"No index type provided.");
+        return nil;
+    }
+    
+    if (![[CDTQIndex validTypes] containsObject:indexType.lowercaseString]) {
+        LogError(@"Invalid index type %@.", indexType);
+        return nil;
+    }
+    
+    if ([indexType.lowercaseString isEqualToString:kCDTQJsonType] && indexSettings) {
+        LogWarn(@"Index type is %@, index settings %@ ignored.", indexType, indexSettings);
+        indexSettings = nil;
+    } else if ([indexType.lowercaseString isEqualToString:kCDTQTextType]) {
+        if (!indexSettings) {
+            indexSettings = @{ kCDTQTextTokenize: kCDTQTextDefaultTokenizer };
+            LogDebug(@"Index type is %@, defaulting settings to %@.", indexType, indexSettings);
+        } else {
+            for (NSString *parameter in [indexSettings allKeys]) {
+                if (![[CDTQIndex validSettings] containsObject:parameter.lowercaseString]) {
+                    LogError(@"Invalid parameter %@ in index settings %@.", parameter,
+                                                                            indexSettings);
+                    return nil;
+                }
+            }
+        }
+    }
+    
+    return [[CDTQIndex alloc] initWithFields:fieldNames
+                                   indexName:indexName
+                                   indexType:indexType
+                               indexSettings:indexSettings];
+}
+
+-(BOOL) compareIndexTypeTo:(NSString *)indexType withIndexSettings:(NSString *)indexSettings
+{
+    if (![self.indexType.lowercaseString isEqualToString:indexType.lowercaseString]) {
+        return NO;
+    }
+    
+    if (!self.indexSettings && !indexSettings) {
+        return YES;
+    } else if (!self.indexSettings || !indexSettings) {
+        return NO;
+    }
+    
+    NSError *error;
+    NSData *settingsData = [indexSettings dataUsingEncoding:NSUTF8StringEncoding];
+    NSDictionary *settingsDict = [NSJSONSerialization JSONObjectWithData:settingsData
+                                                                 options:kNilOptions
+                                                                   error:&error];
+    if (error) {
+        LogError(@"Error processing index settings %@", indexSettings);
+        return NO;
+    }
+    
+    return [self.indexSettings isEqualToDictionary:settingsDict];
+}
+
+-(NSString *) settingsAsJSON {
+    if (!self.indexSettings) {
+        LogWarn(@"Index settings are nil.  Nothing to return.");
+        return nil;
+    }
+    NSError *error;
+    NSData *settingsData = [NSJSONSerialization dataWithJSONObject:self.indexSettings
+                                                           options:kNilOptions
+                                                             error:&error];
+    if (error) {
```

Comment:
Always check the return value of the method rather than checking the `error` parameter to detect an error. There's no guarantee or convention that the error parameter being `nil` or not means the method failed or succeeded.